packages/transformers/assemblyscript/src/ASTransformer.js

```diff
@@ -0,0 +1,144 @@
+// @flow strict-local
+
+import {Transformer} from '@parcel/plugin';
+import ThrowableDiagnostic from '@parcel/diagnostic';
+import {promises as fsp} from 'fs';
+import path from 'path';
+import {tmpdir} from 'os';
+import {Transform} from 'stream';
+
+const MARKER = 'asc:';
+const PREFIX_MATCHER = /^asc:(.+)$/;
+
+// This special import contains the `compileStreaming` polyfill.
+const SPECIAL_IMPORT = '__rollup-plugin-assemblyscript_compileStreaming';
+
+const compileStreamingCode = `
+  export async function compileStreaming(respP) {
+              if('compileStreaming' in WebAssembly) {
+                return WebAssembly.compileStreaming(respP);
+              }
+              return respP
+                .then(resp => resp.arrayBuffer())
+                .then(buffer => WebAssembly.compile(buffer));
+            }
+        `;
+
+export default (new Transformer({
+  async transform({asset, options}) {
+    const asCompiler = await getASCompilerPath(asset.filePath, options);
+
+    const wasmFileName = `${path.basename(asset.filePath)}.wasm`;
+    const folder = tmpdir();
+    const wasmFilePath = path.join(folder, wasmFileName);
+    const sourceMapFileName = wasmFileName + '.map';
+    const sourceMapFilePath = path.join(folder, sourceMapFileName);
+    await asCompiler.ready;
+
+    await new Promise((resolve, reject) => {
+      const errorCollector = streamCollector();
+      const params = [
+        asset.filePath,
+        '-b',
+        wasmFilePath,
+        ...(options.sourceMapURLPattern
+          ? [
+              `--sourceMap=${renderNamePattern(options.sourceMapURLPattern, {
+                name: sourceMapFileName,
+              })}`,
+            ]
+          : []),
+        ...Object.entries(options.compilerOptions).map(([opt, val]) => {
+          if (typeof val === 'boolean') {
+            return `--${opt}`;
+          }
+          return `--${opt}=${val}`;
+        }),
+        ...(options.fileExtension
+          ? [`--extension`, options.fileExtension]
+          : []),
+      ];
+      asCompiler.main(params, {stderr: errorCollector.stream}, async err => {
+        if (err) {
+          errorCollector.stream.end();
+          const stderr = await errorCollector.result;
+          const msg = new TextDecoder().decode(stderr);
+          return reject(msg);
+        }
+        resolve();
+      });
+    });
+
+    // wasm file
+    asset.type = 'wasm';
+    asset.setCode(await fsp.readFile(wasmFilePath));
```

Comment:
Are there downsides to using this variant of the compiler instead? Would be cleaner than writing & reading from a tmpdir and would also work in the browser (from https://github.com/AssemblyScript/assemblyscript/tree/master/cli#api).

![Bildschirmfoto 2021-01-09 um 16 20 52](https://user-images.githubusercontent.com/4586894/104095366-ab73d800-5296-11eb-882d-f2bd19fca89e.png)